src/Init/Data/BitVec/Bitblast.lean

```diff
@@ -429,6 +432,640 @@ theorem shiftLeft_eq_shiftLeftRec (x : BitVec w₁) (y : BitVec w₂) :
+    apply Nat.div_eq_of_lt n.isLt
+}
+
+theorem DivRemInput_implies_udiv_urem
+  (qr : DivRemInput w) (h : DivRemInputProof w w 0 n d qr):
+    n.udiv d = qr.q ∧ n.umod d = qr.r := by
+  apply div_characterized_of_mul_add_toNat
+    (n := n) (d := d) (q := qr.q) (r := qr.r)
+    (h.hd)
+    (h.hrd)
+    (by
+      have hdiv := h.hdiv
+      simp at hdiv
+      omega
+    )
+
+/--
+An input that is given to the shift-subtractor.
+For the input to be legal we need `0 < wn`.
+-/
+def ShiftSubtractInput w := DivRemInput w
+
+/-- Forget that the input is to a shift subtractor, and make it a DivRem input. -/
+def ShiftSubtractInput.toDivRemInput (qr : ShiftSubtractInput w) :
+    DivRemInput w := qr
+
+/-- Forget that the input is to a shift subtractor, and make it a DivRem input. -/
+def ShiftSubtractInput.ofDivRemInput (qr : DivRemInput w) :
+    ShiftSubtractInput w := qr
+
+structure ShiftSubtractInput.lawful (w wr wn : Nat) (n d : BitVec w) (qr : ShiftSubtractInput w)
+  extends DivRemInputProof w wr wn n d qr : Type where
+  hwn_lt : 0 < wn -- we can only call this function legally if we have dividend bits.
+
+/--
+In the shift subtract input, we have one more bit to spare,
+so we do not overflow.
+-/
+def ShiftSubtractInput.wr_add_one_le_w {qr : ShiftSubtractInput w}
+    (h : qr.lawful wr wn n d) : wr + 1 ≤ w := by
+  have hwrn := h.hwrn
+  have hwn_lt := h.hwn_lt
+  omega
+
+def ShiftSubtractInput.wr_lt_w {qr : ShiftSubtractInput w} (h : qr.lawful wr wn n d) : wr < w := by
+  have hwr := qr.wr_add_one_le_w h
+  omega
+
+/--
+In the shift subtract input, we have one more bit to spare,
+so we do not overflow.
+-/
+def ShiftSubtractInput.wr_le_wr_sub_one
+    {qr : ShiftSubtractInput w} (h : qr.lawful wr wn n d) : wr ≤ w - 1 := by
+  have hw := h.hw
+  have hwrn := h.hwrn
+  have hwn_lt := h.hwn_lt
+  omega
+
+/-- If we have extra bits to spare in `n`,
+then the div rem input can be converted into a shift subtract input
+to run a round of the shift subtracter. -/
+def DivRemInputProof.toShiftSubtractInputLawful (qr : DivRemInput w)
+    (h : DivRemInputProof w wr (wn + 1) n d qr) : (ShiftSubtractInput.ofDivRemInput qr).lawful wr (wn + 1) n d := {
+    hwr := h.hwr,
+    hwn := h.hwn,
+    hwrn := by have := h.hwrn; omega,
+    hd := h.hd,
+    hrd := h.hrd,
+    hrwr := h.hrwr,
+    hqwr := h.hqwr,
+    hdiv := h.hdiv,
+    hwn_lt := by omega
+  }
+
+def DivRemInput.wr_eq_w_of_wn_eq_zero (h : DivRemInput w) : DivRemInput w :=
+  {
+    q := h.q,
+    r := h.r,
+  }
+
+def DivRemInputProof.wr_eq_w_of_wn_eq_zero (qr : DivRemInput w)
+    (h : DivRemInputProof w wr 0 n d qr) : DivRemInputProof w w 0 n d qr.wr_eq_w_of_wn_eq_zero :=
+  {
+    hwr := by have := h.hwr; omega,
+    hwn := h.hwn,
+    hwrn := by have := h.hwrn; omega,
+    hd := h.hd,
+    hrd := h.hrd,
+    hrwr := by have := h.hrwr; omega,
+    hqwr := by have := h.hqwr; omega,
+    hdiv := h.hdiv
+  }
+
+def concatBit' (x : BitVec w) (b : Bool) : BitVec w :=
```

Comment:
Why is this labeled with a `'`? Is this not the only version here?

Reply:
We have a `concatBit` over in `BitVecor/Basic.lean` that is of type `BitVec w -> BitVec (w + 1)`. I chose not to use it, because it got me into cast hell (:

Reply:
Ah, it's not really concatenating, it's shifting and concatenating, IIUC. You might want to change the name to reflect that, and include a docstring that emphasizes the difference with `concatBit`

Reply:
Also, this feels like it should be moved to the general definitions file, and its lemmas to `Lemmas`

Reply:
Moved.